support-frontend/test/selenium/util/PostDeployTestUser.scala

```diff
@@ -31,28 +34,56 @@ class PostDeployTestUserContribs(driverConfig: DriverConfig) extends TestUser {
 
 class PostDeployTestUserSubs(driverConfig: DriverConfig) extends TestUser {
 
-  private val client = new OkHttpClient()
-  private val requestToIdapi = new Request.Builder()
-    .url(Config.idapiNewTestUserUrl)
-    .addHeader(Config.idapiClientAccessTokenName, Config.idapiClientAccessTokenSecret)
-    .post(RequestBody.create(MediaType.parse("application/json"), "{}"))
-    .build()
-  private val response = client.newCall(requestToIdapi).execute()
-  private val responseBodyAsJson = parse(response.body().string()).getOrElse(Json.Null)
-  private val cursor = responseBodyAsJson.hcursor
-
   private def addTestUserCookies(testUsername: String) = {
     driverConfig.addCookie(name = "pre-signin-test-user", value = testUsername)
     driverConfig.addCookie(name = "_test_username", value = testUsername, domain = Some(Config.guardianDomain))
     driverConfig.addCookie(name = "_post_deploy_user", value = "true") // This enables the tests to use the mocked payment services
     driverConfig.addCookie(name = "GU_TK", value = "1.1") //To avoid consent banner, which messes with selenium
-    for (n <- 0 until 3) {
-      val cookie_name = cursor.downField("values").downN(n).downField("key").as[String].right.get
-      val cookie_value = cursor.downField("values").downN(n).downField("value").as[String].right.get
-      driverConfig.addCookie(name = cookie_name, value = cookie_value)
+    IdapiRequest.decodeCookies.foreach { cookie =>
+      driverConfig.addCookie(name = cookie.key, value = cookie.value)
     }
   }
 
   val username = "postDeployTestUser"
   addTestUserCookies(username)
 }
+
+case class IdapiCookie(key: String, value: String)
+
+object IdapiCookie {
+  implicit val decoder: Decoder[IdapiCookie] = deriveDecoder[IdapiCookie]
+}
+
+case class IdapiSuccessfulResponse(values: List[IdapiCookie])
+
+object IdapiSuccessfulResponse {
+  implicit val decoder: Decoder[IdapiSuccessfulResponse] = deriveDecoder[IdapiSuccessfulResponse]
+}
+
+object IdapiRequest {
+
+  private val client = new OkHttpClient()
+
+  private def logger = LoggerFactory.getLogger(this.getClass)
+
+  private val request: Request = new Request.Builder()
+    .url(Config.idapiNewTestUserUrl)
+    .addHeader(Config.idapiClientAccessTokenName, Config.idapiClientAccessTokenSecret)
+    .post(RequestBody.create(MediaType.parse("application/json"), "{}"))
+    .build()
+
+  private def getResult: Json = {
+    val response: Response = client.newCall(request).execute()
+    parse(response.body().string()).getOrElse(Json.Null)
+  }
+
+  def decodeCookies: List[IdapiCookie] = {
+    val responseDecoded: Either[DecodingFailure, IdapiSuccessfulResponse] = getResult.as[IdapiSuccessfulResponse]
+    var cookies = List[IdapiCookie]()
+    responseDecoded match {
+      case Left(l) => logger.error(s"Failed to decode IDAPI response: ${l}")
+      case Right(r) => cookies = r.values
+    }
+    cookies
+  }
+}
```

Comment:
The separation looks great, I would just try to standardise on a failure type e.g. `Either[String, A]` and use that throughout.
The purpose of failure types is to consolidate error handling where the error can be dealt with (similar to exceptions)
So where the errors are detected, push them into the `Left` straight away and make any error producing function return an Either.
Then in the code that uses the cookies, we can add the cookies if the call worked, but if it didn't, don't even try to run the test, just abort and fail the test with the error message stored in the Left.
As it is, the test will run and get stuck at the log in screen, which would cause a bit of head scratching to work out actually e.g. identity returned a 5xx error.

Reply:
Makes sense - thank you!

Reply:
@johnduffell - I just read up on Try/Success/Failure, which to my beginner eyes looks like a more readable structure than Either/Left/Right, at least at first sight. Is there a benefit in using the latter?

Reply:
sorry only just saw this.

Either is a generic type on both sides, so you can put whatever you like in either the left or right.  Convention is that the left is failure, but people certainly use Either in other contexts and it's perfectly valid.

Try is specifically relating to exceptions so it's like an Either that the left is hard coded to Throwable.
The naming reflects that specificity, but the downside is you have to wrap any message you have in a Throwable which has a whole load more baggage.

The other option (which I don't recommend here) is to have a custom failure type and you can call it what you want and hard code the failure side to contain a String.

In this case I think Either with a String as the failure i.e. `Either[String,A]` is your best option.  Although the naming is a bit odd, it fits with convention, and you won't have to mess with Throwables.

Reply:
I now see the logic (that Try/Success/Failure involves the use of a Throwable, which has more baggage than just a String). Thank you for explaining!